src/routes/did.rs

```diff
@@ -0,0 +1,192 @@
+
+    hasher.update(format!("{}.{}", header, body));
+    let data_to_verify_hex = hex_encode(hasher.finalize());
+    let data_to_verify = data_to_verify_hex.trim_start_matches("0x").as_bytes();
+
+    let decoded_header = general_purpose::STANDARD.decode(header).map_err(|_|  Error::VerifyJWT)?;
+    let decoded_body = general_purpose::STANDARD.decode(body).map_err(|_|  Error::VerifyJWT)?;
+    let decoded_signature = general_purpose::STANDARD.decode(signature).map_err(|_|  Error::VerifyJWT)?;
+
+    let jwt_header: JWTHeader = serde_json::from_slice(&decoded_header).map_err(|_|  Error::VerifyJWT)?;
+    let jwt_payload: JWTPayload = serde_json::from_slice(&decoded_body).map_err(|_|  Error::VerifyJWT)?;
+
+    let sender = &jwt_payload.iss;
+
+    let endpoint = {
+        let app_state = app_state.read()?;
+        app_state.kilt_endpoint.clone()
+    };
+    let cli = kilt::connect(&endpoint)
+        .await
+        .map_err(|_| Error::CantConnectToBlockchain)?;
+
+    let did_document = get_did_doc(sender, &cli).await?;
+
+    let signed_key = hex_decode(jwt_header.key_uri.trim_start_matches("0x"))
+        .map_err(|_| Error::InvalidDidSignature)?;
+
+    let (_, target_key) = did_document
+        .public_keys
+        .0
+        .iter()
+        .find(|(key, _)| key.as_bytes() == signed_key)
+        .ok_or(Error::InvalidDidSignature)?;
+
+    let public_key = match &target_key.key {
+        DidPublicKey::PublicVerificationKey(DidVerificationKey::Sr25519(pk)) => {
+            Ok(sp_runtime::AccountId32::from(pk.0))
+        }
+        DidPublicKey::PublicVerificationKey(DidVerificationKey::Ed25519(pk)) => {
+            Ok(sp_runtime::AccountId32::from(pk.0))
+        }
+        _ => Err(Error::InvalidDidSignature),
+    }?;
+
+
+    let signature_string = String::from_utf8(decoded_signature).map_err(|_| Error::VerifyJWT)?;
+    let trimed_signature = hex_decode(signature_string.trim_start_matches("0x")).map_err(|_| Error::VerifyJWT)?;
+
+    let valid = {
+        if let Ok(signature) =
+            sp_runtime::MultiSignature::decode(&mut IoReader(trimed_signature.as_slice()))
+        {
+            signature.verify(data_to_verify, &public_key)
+        } else if let Ok(signature) =
+            sp_core::sr25519::Signature::decode(&mut IoReader(trimed_signature.as_slice() ))
+        {
+            signature.verify(
+                data_to_verify,
+                &sp_core::sr25519::Public(public_key.into()),
+            )
+        } else if let Ok(signature) =
+            sp_core::ed25519::Signature::decode(&mut IoReader(trimed_signature.as_slice()))
+        {
+            signature.verify(
+                data_to_verify,
+                &sp_core::ed25519::Public(public_key.into()),
+            )
+        } else {
+            false
+        }
+    };
+
+    if !valid {
+        return Err(Error::InvalidDidSignature);
+    }
+
+    let w3n = kilt::get_w3n(&jwt_payload.iss, &cli).await.unwrap_or("".into());
```

Comment:
We need to check if payload.iss is actually the DID that signed the JWT. Otherwise I could just put your DID there

Reply:
I do but looks like it was not declarative enough. Is this better? [aa52209](https://github.com/KILTprotocol/opendid/pull/28/commits/aa52209ad8f05798dacd7ebdc852cf69777cc910)